examples/dpo_demo_gemma3.ipynb

```diff
@@ -0,0 +1,849 @@
+    "  )\n",
+    "  nnx.display(gemma3)"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "print(\"\\n--- HBM Usage AFTER Model Load ---\")\n",
+    "show_hbm_usage()"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "\n",
+    "\n",
+    "gemma_tokenizer = data_lib.Gemma3Tokenizer()\n",
+    "#from transformers import AutoTokenizer\n",
+    "\n",
+    "#gemma_tokenizer = AutoTokenizer.from_pretrained(\"google/gemma-3-1b-it\")\n",
+    "sampler = sampler_lib.Sampler(\n",
+    "    transformer=gemma3,\n",
+    "    tokenizer=gemma_tokenizer,\n",
+    "    cache_config=sampler_lib.CacheConfig(\n",
+    "        cache_size=MAX_PROMPT_LENGTH + TOTAL_GENERATION_STEPS + 256,\n",
+    "        num_layers=model_config.num_layers,\n",
+    "        num_kv_heads=model_config.num_kv_heads,\n",
+    "        head_dim=model_config.head_dim,\n",
+    "    ),\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def get_lora_model(base_model, mesh):\n",
+    "  lora_provider = qwix.LoraProvider(\n",
+    "      module_path=(\n",
+    "          \".*q_einsum|.*kv_einsum|.*gate_proj|.*down_proj|.*up_proj|\"\n",
+    "          \".*attn_vec_einsum\"\n",
+    "      ),\n",
+    "      rank=RANK,\n",
+    "      alpha=ALPHA,\n",
+    "      #weight_qtype=\"nf4\",\n",
+    "      #tile_size=4,\n",
+    "  )\n",
+    "\n",
+    "  model_input = base_model.get_model_input()\n",
+    "  lora_model = qwix.apply_lora_to_model(\n",
+    "      base_model, lora_provider, **model_input\n",
+    "  )\n",
+    "\n",
+    "  with mesh:\n",
+    "    state = nnx.state(lora_model)\n",
+    "    pspecs = nnx.get_partition_spec(state)\n",
+    "    sharded_state = jax.lax.with_sharding_constraint(state, pspecs)\n",
+    "    nnx.update(lora_model, sharded_state)\n",
+    "\n",
+    "  return lora_model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Policy model\n",
+    "lora_gemma = get_lora_model(gemma3, mesh=mesh)\n",
+    "nnx.display(lora_gemma)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "id": "UNwb0GKuwFYb"
+   },
+   "source": [
+    "Load evaluation data and evaluate the reference model"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "eval_dataset = load_dataset(\"gsm8k\", \"main\", split=\"test\").select(range(NUM_TEST_BATCHES))"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "reasoning_start = \"<reasoning>\"\n",
+    "reasoning_end = \"</reasoning>\"\n",
+    "solution_start = \"<answer>\"\n",
+    "solution_end = \"</answer>\"\n",
```

Comment:
Confused why it's not showing up in the notebook?

<img width="149" height="68" alt="image" src="https://github.com/user-attachments/assets/6ee2f73f-28d7-4e37-99c6-0faed476c81b" />